guides/_posts/2015-01-06-constraints.html

```diff
@@ -0,0 +1,839 @@
+For more than a decade, no one seemed to take notice outside academia... Until Apple implemented Cassowary & Greg's pioneering concepts in its new
+AutoLayout engine with the launch of OS X Lion. Despite the evolutionary leap for app developers, web designer's have had to settle
+with [float-based](http://www.sitepoint.com/give-floats-the-flick-in-css-layouts/) & table-based layouts that have remained unimproved to this day.
+
+> "We should contemplate how very, very far behind the web platform is in making it delightful to build the sorts of things that are work-a-day in native environments." <a href="http://infrequently.org/2012/02/misdirection/">Alex Russell</a>
+
+The foundation of GSS is a modernized implementation of CCSS.  From CCSS primitives other, more exotic layout API's can and are accomplished.
+
+<a name="basicconstraint"></a>
+## Constraints, the basics
+
+Constraints express [relationships between variables](#constraintsaretwoways) that [may or may not hold](#mayofmaynothold). You can constrain any
+numerical style property of an element, not just its position & size.
+
+For example, if I want all paragraph tags to have line-height greater than 16px and less than 1/12th the size of the window:
+
+{% highlight css %}
+
+p[line-height] >= 16;
+p[line-height] <= ::window[height] / 12;
+
+{% endhighlight %}
+
+### Constraints declaration
+
+Let's go through the syntax for declaring a constraint:
+
+`p[line-height] >= 16;`
+
++ `p` is the selector. This constraint will apply to every `p` tag in the page.
++ `[]` is the style property selector.
++ `line-height` is the style property for which a value will be calculated by GSS.
++ `>=` defines an inequality constraint. You can declare inequality and equality constraint.
++ `16` is the numerical value for the constraint in pixel unit. The default unit in GSS are pixels.
+
+<a href="http://codepen.io/cbchouinard/pen/OPmdVL" target="_blank">Live example</a>
+
+<a name="constraintsaretwoways"></a>
+### Constraints are 2-ways
+Do not confuse an equality constraint with your everyday variable assignment. For example, in a vanilla programming paradigm:
+
+{% highlight javascript %}
+
+// vanilla programming
+
+x = y;
+y = 10
+x = 100;
+
+// x => 100, y => 10
+
+{% endhighlight %}
+
+x will be 100, and y will be 10.  With constraint programming, equality constraints are two-way, so:
+
+{% highlight javascript %}
+
+// constraint programming
+
+x == y;
+x == 100;
+y >= 10;
+
+// x => 100, y => 100
+
+{% endhighlight %}
+
+x and y have a constraint to be equal, so x and y will be 100.
+
+Constraints are incrementally added to a `solver`, then the solver computes a feasible & optimal solution to all the constraints.
+With constraint programming you input the *what* you want to happen, not *how* it will happen. With procedural programming,
+you focus on the *how*.  This makes constraints a perfect fit for empowering declarative languages like CSS.
```

Comment:
> Constraint programming takes *what* should happen as input, not *how* it will happen. Procedural programming focuses on the *how*.

(removes "you")